crates/turborepo-lib/src/rewrite_json.rs

```diff
@@ -0,0 +1,411 @@
+    NoRoot,
+}
+/**
+ * When generating replacement content it's one of two things.
+ */
+enum GenerateType {
+    Object,
+    Member,
+}
+
+/**
+ * Range stores the primitive AST token details required for document
+ * mutation.
+ */
+struct Range {
+    pub start: usize,
+    pub end: usize,
+    pub replacement_char: String,
+}
+
+/**
+ * Given a JSONC document, an object traversal path, and a _pre-serialized_
+ * JSON value `set_path` will return a minimally-mutated JSONC document with
+ * the path specified set to the JSON value.
+ *
+ * - If the path exists, it will clobber the existing contents.
+ * - If the path does not exist, it will synthesize object members and
+ *   objects to ensure it does exist.
+ *
+ * In the event that the key appears multiple times in the document the last
+ * instance of the key will be updated and earlier instances will be
+ * disregarded.
+ */
+pub fn set_path(
+    json_document_string: &str,
+    path: Vec<&str>,
+    json_value: &str,
+) -> Result<String, RewriteError> {
+    let root = get_root(json_document_string)?;
+
+    // Find the token we'll be modifying and its path from the root.
+    let (closest_path, closest_node) = get_closest_node(&root, &path, vec![]);
+
+    // Pull the token metadata off of the token.
+    let (property_count, range): (usize, jsonc_parser::common::Range) = match closest_node {
+        jsonc_parser::ast::Value::Object(literal) => (literal.properties.len(), literal.range),
+        jsonc_parser::ast::Value::StringLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NumberLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::BooleanLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::Array(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NullKeyword(literal) => (0, literal.range),
+    };
+
+    // Figure out what we should be generating:
+    // - An object to be assigned to an existing member. ("object")
+    // - A member to add to an existing object. ("member")
+    let generate_type: GenerateType = if !closest_path.is_empty() {
+        GenerateType::Object
+    } else {
+        GenerateType::Member
+    };
+
+    // Identify the token replacement metadata: start, end, and possible trailing
+    // join character
+    let separator;
+    let start;
+    let end;
+
+    match generate_type {
+        GenerateType::Object => {
+            separator = "";
+            start = range.start;
+            end = range.end;
+        }
+        GenerateType::Member => {
+            if property_count > 0 {
+                separator = ",";
+            } else {
+                separator = ""
+            }
+
+            start = range.start + 1;
+            end = range.start + 1;
+        }
+    }
+
+    // Generate the serialized JSON to insert into the document.
+    // We synthesize objects for missing path segments.
+    let missing_path_segments = path[closest_path.len()..].to_vec();
```

Comment:
Unnecessary allocation
```suggestion
    let missing_path_segments = &path[closest_path.len()..];
```